File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    maxDepth(root) {

        if(!root) return 0
        let ld = this.maxDepth(root.left)
        let rd = this.maxDepth(root.right)
        let depth = 1 + (ld > rd ? ld : rd)

        return depth

    }

}

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-0.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public TreeNode BuildTree(int[] preorder, int[] inorder) {


        TreeNode traverse(ReadOnlySpan<int> preorder,ReadOnlySpan<int> inorder){
            if(preorder.Length == 0) return null;

            var root = new TreeNode(preorder[0]);

            var i = inorder.IndexOf(root.val);

            root.left = traverse(preorder.Slice(1,i), inorder.Slice(0,i));
            root.right = traverse(preorder.Slice(i+1), inorder.Slice(i+1));

            return root;
        }

        return traverse(preorder.AsSpan(),inorder.AsSpan());
        
    }
}
